pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -111,54 +162,46 @@ func newCDCMixedVersionTester(
 	c.Put(ctx, t.DeprecatedWorkload(), "./workload", lastNode)
 
 	return cdcMixedVersionTester{
-		ctx:               ctx,
-		crdbNodes:         crdbNodes,
-		workloadNodes:     lastNode,
-		kafkaNodes:        lastNode,
-		monitor:           c.NewMonitor(ctx, crdbNodes),
-		validatorFinished: make(chan struct{}),
+		ctx:           ctx,
+		crdbNodes:     crdbNodes,
+		workloadNodes: lastNode,
+		kafkaNodes:    lastNode,
+		workloadInit:  make(chan struct{}),
 	}
 }
 
 // StartKafka will install and start Kafka on the configured node. It
-// will also create the topic where changefeed events will be sent.
-func (cmvt *cdcMixedVersionTester) StartKafka(t test.Test, c cluster.Cluster) {
+// will also create the topic where changefeed events will be sent and a consumer.
+// Returns a function to clean up the consumer and kafka resources.
+func (cmvt *cdcMixedVersionTester) StartKafka(t test.Test, c cluster.Cluster) (cleanup func()) {
 	t.Status("starting Kafka node")
-	cmvt.kafka, cmvt.cleanup = setupKafka(cmvt.ctx, t, c, cmvt.kafkaNodes)
-	if err := cmvt.kafka.createTopic(cmvt.ctx, targetTable); err != nil {
+
+	manager, tearDown := setupKafka(cmvt.ctx, t, c, cmvt.kafkaNodes)
+	if err := manager.createTopic(cmvt.ctx, targetTable); err != nil {
 		t.Fatal(err)
 	}
-}
+	cmvt.kafka.manager = manager
 
-// Cleanup is supposed to be called at the end of tests that use
-// `cdcMixedVersionTester`
-func (cmvt *cdcMixedVersionTester) Cleanup() {
-	if cmvt.cleanup != nil {
-		cmvt.cleanup()
+	consumer, err := cmvt.kafka.manager.newConsumer(cmvt.ctx, targetTable)
+	if err != nil {
+		t.Fatal(err)
 	}
-}
 
-// installAndStartWorkload starts a bank workload asynchronously
-func (cmvt *cdcMixedVersionTester) installAndStartWorkload() versionStep {
-	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
-		t.Status("installing and running workload")
-		u.c.Run(ctx, cmvt.workloadNodes, "./workload init bank {pgurl:1}")
-		cmvt.monitor.Go(func(ctx context.Context) error {
-			return u.c.RunE(
-				ctx,
-				cmvt.workloadNodes,
-				fmt.Sprintf("./workload run bank {pgurl%s} --max-rate=10 --tolerate-errors",
-					cmvt.crdbNodes),
-			)
-		})
+	cmvt.kafka.manager = manager
+	cmvt.kafka.buf = make([]*sarama.ConsumerMessage, 0, kafkaBufferMessageSize)
+	cmvt.kafka.consumer = consumer
+
+	return func() {
+		cmvt.kafka.consumer.Close()
+		tearDown()
 	}
 }
 
-// waitForResolvedTimestamps waits for the underlying CDC verifier to
-// resolve `resolvedTimestampsPerState`
-func (cmvt *cdcMixedVersionTester) waitForResolvedTimestamps() versionStep {
-	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
-		t.Status(fmt.Sprintf("waiting for %d resolved timestamps", resolvedTimestampsPerState))
+// waitForResolvedTimestamps waits for a few resolved timestamps to assert
+// that the changefeed is running.
+func (cmvt *cdcMixedVersionTester) waitForResolvedTimestamps() mixedversion.UserFunc {
```

Comment:
See comment about making these implement `userFunc` directly -- why not `waitForResolvedTimestamps(ctx context.Context, l *logger.Logger, r *rand.Rand, h *mixedversion.Helper)`?